src/generic/FreeAssAlgebraGroebner.jl

```diff
@@ -0,0 +1,640 @@
+            del_counter += 1
+            delete!(all_obstructions, obstr_pair[1])
+        end
+    end
+    # TODO case 4e from Thm 4.1 in Kreuzer Xiu
+end
+
+function get_obstructions(g::Vector{FreeAssAlgElem{T}}) where {T}
+    s = length(g)
+    result = PriorityQueue{Obstruction{T},FreeAssAlgElem{T}}()
+    for i in 1:s, j in 1:i
+        if i == j
+            obs = obstructions(_leading_word(g[i]))
+        else
+            obs = obstructions(_leading_word(g[i]), _leading_word(g[j]))
+        end
+        for o in obs
+            triple = ObstructionTriple{T}(g[i], g[j], o, i, j)
+            enqueue!(result, triple, common_multiple_leading_term(triple))
+        end
+    end
+    # TODO maybe here some redundancies can be removed too, check Kreuzer Xiu
+    return result
+end
+
+
+function add_obstructions!(
+    obstruction_queue::PriorityQueue{Obstruction{T},FreeAssAlgElem{T}},
+    g::Vector{FreeAssAlgElem{T}},
+) where {T}
+    s = length(g)
+    for i in 1:s
+        if i == s
+            obs = obstructions(_leading_word(g[i]))
+        else
+            obs = obstructions(_leading_word(g[i]), _leading_word(g[s]))
+        end
+        for o in obs
+            triple = ObstructionTriple{T}(g[i], g[s], o, i, s)
+            enqueue!(obstruction_queue, triple, common_multiple_leading_term(triple))
+        end
+    end
+    #remove_redundancies!(obstruction_queue, s, g[s]) #TODO too slow in practice
+end
+
+
+function groebner_basis_buchberger(
+    g::Vector{FreeAssAlgElem{T}},
+    reduction_bound::Int = typemax(Int),
+    remove_redundancies::Bool = false
+) where {T<:FieldElement}
+    g = copy(g)
+    #   interreduce!(g) # on some small examples, this increases running time, so it might not be optimal to use this here
+    nonzero_reductions = 0
+    # compute the aho corasick automaton
+    # to make normal form computation more efficient
+    aut = AhoCorasickAutomaton([g_i.exps[1] for g_i in g])
+    # step 1
+    obstruction_queue = get_obstructions(g)
+    while !isempty(obstruction_queue)
+        obstruction = dequeue!(obstruction_queue)
+        # step3
+        S = s_polynomial(obstruction)
+        Sp = normal_form(S, g, aut) # or normal_form_weak
+        if iszero(Sp)
+            continue
+        end
+        nonzero_reductions += 1
+        # step4
+        push!(g, Sp)
+        insert_keyword!(aut, Sp.exps[1], length(g))
+        if nonzero_reductions >= reduction_bound
+            return g
+        end
+        add_obstructions!(obstruction_queue, g)
+        if remove_redundancies
+            remove_redundancies!(obstruction_queue, length(g), g[length(g)])
+        end
+    end
+    return g
+end
+
+@doc """
+    groebner_basis(g::Vector{FreeAssAlgElem{T}}, reduction_bound::Int = typemax(Int), remove_redundancies::Bool = false)
+
+Compute a groebner basis for the ideal spanned by g. Stop when `reduction_bound` many
+non-zero entries have been added to the groebner basis.
```

Comment:
```suggestion
non-zero entries have been added to the Groebner basis.
```
But in that case the result is (in general) not a Groebner basis? I think this should be pointed out